io/src/main/scala/sbt/io/IO.scala

```diff
@@ -454,14 +455,24 @@ object IO {
 
   /** Deletes `file`, recursively if it is a directory. */
   def delete(file: File): Unit = {
-    translate("Error deleting file " + file + ": ") {
-      val deleted = file.delete()
-      if (!deleted && file.isDirectory) {
-        delete(listFiles(file))
-        file.delete
-        ()
+    object deleter extends SimpleFileVisitor[Path] {
+      override def visitFile(file: Path, attr: BasicFileAttributes): FileVisitResult = {
+        translate("Error deleting file " + file.toFile + ": "){
+          Files.delete(file)
+        }
+        FileVisitResult.CONTINUE
+      }
+
+      override def postVisitDirectory(dir: Path, e: IOException): FileVisitResult = {
+        if (e eq null) {
+          translate("Error deleting file " + dir.toFile + ": ") {
```

Comment:
I think we should wrap the whole `walkFileTree`, rather than specific code branches of the file visiting.

Reply:
I didn't want to change the error message, and the file that failed is not visible directly, so the message would be platform specific, not sure how much sbt is support to insulate you from this detail

I did consider a last file var, and that would be slightly faster, but didn't want to enrage the functional reviewers :-)

It would seem OK to wrap the operation though - happy to take steer - this is your library. I am just doing a fly-by fix

Reply:
I guess there might be benefit in knowing _both_:
1. the given `file` argument passed into `IO.delete(...)` by the user and
2. the inner `file`/`dir` that's causing the exception.

WDYT?

As per `var`, I am personally fine with the usage especially when interacting with Java, when it's confined to a method so it won't cause race conditions, and when it has performance upside.